### PR TITLE
Various cleanups

### PR DESCRIPTION
**Pull request by rbtcollins**
Hopefully this all makes sense. I'm not 100% sure my raw pointer replacement is actually safe-and-sane, but AFAICT it is.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -596,11 +596,11 @@ impl NetmapDescriptor {
         return None;
     }
 
-    #[allow(dead_code)]
     pub fn get_fd(&self) -> i32 {
         self.desc.get().fd
     }
 
+    #[deprecated]
```
Comment (polachok) [inline src/lib.rs:574]: I disagree with your reasoning here. For example, in my project I give a netmap descriptor for a ring to each thread and call poll(). But yes, it kinda feels out of place here.

Comment (rbtcollins) [inline src/lib.rs:574]: Sure - you can definitely do that too - still can, but I'd put the method on your thread object perhaps? Anyhow - do you want me to back this part of the change out?

**Comment by polachok** [comment]
About Unique: I don't want to lock this crate to nightly compiler unless absolutely necessary. And I don't quite see how it is necessary in that case.

**Comment by rbtcollins** [comment]
I can probably implement it by-hand, sure. I'll rewrite that stage once I've any other feedback you have.

File: src/lib.rs [hunk 2]
```diff
@@ -445,17 +439,24 @@ impl NetmapDescriptor {
...
-                                 parent.raw)
-        };
-        if netmap_desc == ptr::null_mut() {
+        NetmapDescriptor::_prim_new(&netmap_iface,
+                                    &base_nmd,
+                                    netmap_user::NM_OPEN_NO_MMAP as u64,
+                                    parent.raw)
+    }
+
+    fn _prim_new(netmap_iface: &CString,
```
Comment (polachok) [inline src/lib.rs]: I'm not a fan of the front underscore.

**Comment by polachok** [comment]
I left a comment about new helper, with it fixed and Unique removed I think I can accept it.

**Comment by rbtcollins** [comment]
Sorry, I don't see a comment about new helper ? I saw one about poll, and
Unique.

**Comment by polachok** [comment]
https://github.com/polachok/netmap-rs/pull/2/commits/f1012fc2f3efb8c8712b18237bc7e74be430a89c here (in case you still don't see it I suggest removing front underscore)

**Comment by rbtcollins** [comment]
underscore removed, use of Unique removed.

**Comment by polachok** [comment]
Why do we need WrappedDrop?

**Comment by rbtcollins** [comment]
I could write it without it, but it seemed cleaner to me to keep the two layers separate - really netmap_sys shouldn't be handling out a C pointer ever, given that netmap doesn't keep any hidden references or close the pointer except in nm_close. So at least in my head netmap is partly working around netmap_sys deficiencies, and partly providing clean rust access to the functionality.

**Comment by polachok** [comment]
> really netmap_sys shouldn't be handling out a C pointer ever

Why? In my understanding, that's the whole point of *-sys crates existence  They just provide whatever the interface underlying C API provides without any abstraction. And then you build a higher-level interface on top of it (netmap-rs).

I don't see value in adding another layer of indirection inside this crate.

**Comment by rbtcollins** [comment]
I think you know rust idioms better than I, so tell me what you'd like and I"ll do it :). In other places I'd try to make any one class do just one thing, WrappedDrop let me pull out just one thing (maintain the C handle correctly and safely) from the rest of NetmapDescriptor which is transforming the API to a safer rust one.

**Comment by polachok** [comment]
Ok, I'm for dropping :) the drop wrapper.

Btw you could do it like this and avoid Box'ing

``` diff
diff --git a/src/lib.rs b/src/lib.rs
index 2412df8..8af863a 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -379,12 +379,12 @@ impl<'a> Drop for TxSlotIter<'a> {
 /// Wrap a C pointer with safe semantics.
 struct WrappedDrop<T> {
     pointer: *mut T,
-    dropper: Box<Fn(&mut T)>,
+    dropper: fn(&mut T),
 }

 impl<T> WrappedDrop<T> {
     /// Construct a WrappedDrop from a mutable pointer.
-    pub fn new(ptr: *mut T, dropper: Box<Fn(&mut T)>) -> Option<WrappedDrop<T>> {
+    pub fn new(ptr: *mut T, dropper: fn(&mut T)) -> Option<WrappedDrop<T>> {
         if ptr.is_null() {
             return None;
         }
@@ -408,7 +408,7 @@ impl<T> WrappedDrop<T> {

 impl<T> Drop for WrappedDrop<T> {
     fn drop(self: &mut WrappedDrop<T>) {
-        (*self.dropper)(unsafe { &mut *self.pointer });
+        (self.dropper)(unsafe { &mut *self.pointer });
     }
 }

@@ -487,7 +487,7 @@ impl NetmapDescriptor {
         if let Some(nd) = unsafe {
             netmap_user::nm_open(netmap_iface.as_ptr(), base_nmd, flags, parent).as_mut()
         } {
-            if let Some(desc) = WrappedDrop::new(nd, Box::new(drop_nm_desc)) {
+            if let Some(desc) = WrappedDrop::new(nd, drop_nm_desc) {
                 Ok(NetmapDescriptor { desc: desc })
             } else {
                 Err(NetmapError::new(format!("Couldn't wrap netmap descriptor")))
```

**Comment by rbtcollins** [comment]
I'll remove WrappedDrop. BTW I tried something like that and hit lifetime issues which is why I boxed it up :/.

**Pull request merged**